go/mcap/indexed_message_iterator.go

```diff
@@ -237,13 +238,54 @@ func (it *indexedMessageIterator) loadChunk(chunkIndex *ChunkIndex) error {
 	return nil
 }
 
+func readRecord(r io.Reader) (TokenType, []byte, error) {
+	buf := make([]byte, 9)
+	_, err := io.ReadFull(r, buf)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to read record header: %w", err)
+	}
+	tokenType := TokenType(buf[0])
+	recordLen := binary.LittleEndian.Uint64(buf[1:])
+	record := make([]byte, recordLen)
+	_, err = io.ReadFull(r, record)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to read record: %w", err)
+	}
+	return tokenType, record, nil
+}
+
 func (it *indexedMessageIterator) Next(p []byte) (*Schema, *Channel, *Message, error) {
 	if !it.hasReadSummarySection {
 		err := it.parseSummarySection()
 		if err != nil {
 			return nil, nil, nil, err
 		}
+		// take care of the metadata here
+		if it.metadataCallback != nil {
+			for _, idx := range it.metadataIndexes {
```

Comment:
Would there be any advantage to trying to insert the metadata between messages based on where it appeared in the original file order? I guess since metadata records don't have a timestamp, that kind of implies there's no expected ordering with respect to messages. Seems like maybe this is at least worth a doc comment pointing out that an indexed message iterator will read all metadata records up front (if indexed).

Reply:
yeah - not only do they not have timestamps, but they don't get interleaved with messages in the file, since the metadatas are written outside the chunks.

I think a bigger concern is whether metadata records are ordered with respect to each other. If you have collisions between merged metadata records, which should take precedence? We can make this deterministic but I'm not sure if we can make a principled decision.